rust/src/mqtt/parser.rs

```diff
@@ -0,0 +1,631 @@
+       do_parse!(
+           reason_code: be_u8
+           >> properties: call!(parse_properties, true)
+           >>  (
+                 MQTTAuthData {
+                   reason_code: reason_code,
+                   properties: properties,
+                 }
+               )
+       ));
+
+pub fn parse_message(input: &[u8], protocol_version: u8) -> IResult<&[u8], MQTTMessage> {
+    // Parse the fixed header first. This is identical across versions and can
+    // be between 2 and 5 bytes long.
+    match parse_fixed_header(input) {
+        Ok((fullrem, header)) => {
+            let len = header.remaining_length as usize;
+            // This is the length of the fixed header that we need to skip
+            // before returning the remainder. It is the sum of the length
+            // of the flag byte (1) and the length of the message length
+            // varint.
+            let skiplen = 1 + varint_length(len);
+            let message_type = header.message_type;
+            if fullrem.len() < len {
+                return Err(Err::Incomplete(Needed::Size(len - fullrem.len())));
+            }
+            // We reslice the remainder into the portion that we are interested
+            // in, according to the length we just parsed. This helps with the
+            // complete! parsers, where we would otherwise need to keep track
+            // of the already parsed length.
+            let rem = &fullrem[..len];
+            match message_type {
+                1 => match parse_connect(rem) {
+                    Ok((_rem, conn)) => {
+                        let msg = MQTTMessage {
+                            header: header,
+                            op: MQTTOperation::CONNECT(conn),
+                        };
+                        Ok((&input[skiplen+len..], msg))
+                    }
+                    Err(e) => Err(e),
+                },
+                2 => match parse_connack(rem, protocol_version) {
+                    Ok((_rem, connack)) => {
+                        let msg = MQTTMessage {
+                            header: header,
+                            op: MQTTOperation::CONNACK(connack),
+                        };
+                        Ok((&input[skiplen+len..], msg))
+                    }
+                    Err(e) => Err(e),
+                },
+                3 => match parse_publish(rem, protocol_version, header.qos_level > 0) {
+                    Ok((_rem, publish)) => {
+                        let msg = MQTTMessage {
+                            header: header,
+                            op: MQTTOperation::PUBLISH(publish),
+                        };
+                        Ok((&input[skiplen+len..], msg))
+                    }
+                    Err(e) => Err(e),
+                },
+                4..=7 => {
+                    match parse_msgidonly(rem, len, protocol_version) {
+                        Ok((_rem, msgidonly)) => {
```

Comment:
`rem` and `len` are redundant, can we just use `rem` ?

Reply:
Done.